### PR TITLE
users service: add sessions

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Added session store to the user service. This migration will continue with adding all session methods to the service. ~~I want to get early feedback before adding remaining methods in the `app/session.go`.~~

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: services/users/users.go [hunk 1]
```diff
@@ -6,28 +6,51 @@ package users
...
 type UserCreateOptions struct {
 	Guest      bool
 	FromImport bool
 }
 
-func New(s store.UserStore, cfgFn func() *model.Config) *UserService {
-	return &UserService{
-		store:  s,
-		config: cfgFn,
+func New(s store.UserStore, ss store.SessionStore, c einterfaces.ClusterInterface, m einterfaces.MetricsInterface, cfgFn func() *model.Config) (*UserService, error) {
```
Comment (agnivade) [inline services/users/users.go:33]: This is reaching the threshold of too many params. Might be good time to have an initializer struct / functional options. 0/5.

File: services/users/users.go [hunk 2]
```diff
@@ -6,28 +6,51 @@ package users
...
+	cacheProvider := cache.NewProvider()
+	if err := cacheProvider.Connect(); err != nil {
+		return nil, fmt.Errorf("could not create cache provider: %w", err)
+	}
+
+	sessionCache, err := cacheProvider.NewCache(&cache.CacheOptions{
+		Size:           model.SESSION_CACHE_SIZE,
```
Comment (agnivade) [inline services/users/users.go:40]: Should `SESSION_CACHE_SIZE` be moved to this package?

File: services/users/session.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+func (us *UserService) SessionCacheLength() int {
+	if l, err := us.sessionCache.Len(); err == nil {
+		return l
+	}
+	return 0
+}
+
+func (us *UserService) ClearSessionCacheForUserSkipClusterSend(userID string) {
```
Comment (agnivade) [inline services/users/session.go:64]: I'd like to change the naming from "skipclustersend" to "local". Thinking of something like this:
```
ClearUserSessionCache
ClearUserSessionCacheLocal
ClearAllUsersSessionCache
ClearAllUsersSessionCacheLocal
```

**Review by agnivade — COMMENTED** [review]
Nice work. Really like where this is going.

**Review by streamer45 — COMMENTED** [review]
Agree with @agnivade's points. I've got just a couple of additional comments from an _outsider_'s point of view:

- I've honestly got confused by the package name `users`. I feel that may hinder a bit the readability because most of the time you expect that to identify an object of some kind, not a namespace.
- What was the reasoning for keeping some functions outside the service methods pool? For example, couldn't the `users.ReturnSessionToPool()` more easily become `userService.ReturnSessionToPool()`? Especially since `GetSession()` is.

File: services/users/users.go (hunk 1 above)
Comment (isacikgoz) [reply services/users/users.go:33]: I liked the idea 👍 Also, what about having both? An initializer for madatory fields and functional arguments for optional parameters?

File: services/users/users.go (hunk 2 above)
Comment (isacikgoz) [reply services/users/users.go:40]: Checked and this is also used in `store` package [here](https://github.com/mattermost/mattermost-server/blob/46649292f8149dbfb4e29c55a3f40140717cc5b6/store/sqlstore/channel_store.go#L30)

**Comment by isacikgoz** [comment]
@streamer45 great points, For the package naming; this was falling under `services/` and I didn't want to have a package name too long. Also other services under the same directory are mostly named w/o service suffix. But I'm open to new ideas and suggestions. 
For the second point, it was same in the `app` package, the reasoning for me to not exporting the service outside from the `app` package therefore I didn't changed it. But we can add `app.App` methods so that we can attach these functions to the service as methods.

**Comment by streamer45** [comment]
> @streamer45 great points, For the package naming; this was falling under `services/` and I didn't want to have a package name too long. Also other services under the same directory are mostly named w/o service suffix. But I'm open to new ideas and suggestions.

I think I'd at least made it singular (`user`). Also, `userservice` is not that long and may better convey the actual meaning of the package.

File: services/users/users.go (hunk 1 above)
Comment (agnivade) [reply services/users/users.go:33]: Feels a bit odd, but no objections from me. :+1:

File: services/users/users.go (hunk 2 above)
Comment (agnivade) [reply services/users/users.go:40]: Hmm .. I'd say those things should really be unexported constants under the store package. But this is getting out of scope. Fine to leave it as-is.

**Comment by agnivade** [comment]
>  Also, userservice is not that long and may better convey the actual meaning of the package.

Hmm .. this maybe nitpicking a bit, but the package is under `services`. So any package under that is meant to be a service. The import statement reads like `import "bla/bla/services/users"` which I think is fine I think.

But I guess the issue lies is when you initialize it with `users.New()`. Wondering if an import aliasing would help here? That way we can keep the package name short, but still maintain readability.

`import userservice "bla/bla/services/users"`

Apologies if this sounded a bit nitpicky :stuck_out_tongue_closed_eyes: . Just 1/5 here.

**Comment by streamer45** [comment]
Yes, my main concern is the ambiguity of the package name if taken as-is, especially out of the "import" context. But nothing blocking really, just some feedback.

**Comment by isacikgoz** [comment]
This is started to get bigger, let's merge this when the reviews are ok and proceed with the next one.

File: services/users/users.go [hunk 4]
```diff
@@ -5,29 +5,77 @@ package users
...
 	Guest      bool
 	FromImport bool
 }
 
-func New(s store.UserStore, cfgFn func() *model.Config) *UserService {
-	return &UserService{
-		store:  s,
-		config: cfgFn,
+// ServiceInitializer is used to initialize the UserService.
+type ServiceInitializer struct {
```
Comment (agnivade) [inline services/users/users.go:37]: Maybe `ServiceConfig`? 0/5

**Review by agnivade — APPROVED** [review]
Great work chipping away on this. :+1:

**Review by streamer45 — APPROVED** [review]

File: services/users/users.go (hunk 4 above)
Comment (isacikgoz) [reply services/users/users.go:37]: Sure, I'll include that in the next PR 👍

**Pull request merged**